Add basic access control

andrew-gardener opened the pull request:
- added `enrollment_type` to user table (currently only stores `StudentEnrollment`, `TaEnrollment`, and `TeacherEnrollment` enrolments)
- modified user cron query to fetch student, ta, and instructor enrollments (instead of just students) as well as passing along `enrollment_type` info
- added rules (django-rules) for basic access control
	- Note missing 403 error handling to api calls from front end (didn't add due to react frontend rewrite) as well as a 403.html template

Related to #314

jonespm commented on dashboard/cron.py:
```diff
@@ -128,25 +128,29 @@ def update_user(self):
+                        join pseudonym_dim p ON p.user_id = u.id
+                        join enrollment e ON u.id = e.user_id
+                        where p.workflow_state = 'active'
+                        and p.authentication_provider_id = '{settings.CANVAS_PRIMARY_AUTH_PROVIDER_ID}'
```
This value authentication_provider_id looks to be "null" for all of our values in the table. Joining this this against pseudonym_dim would have to be changed to allow for "is null" or be an optional condition.

jonespm commented:
This needs to be updated to support the models change, init.sql has been removed.

pushyamig commented on dashboard/cron.py:
```diff
@@ -128,25 +128,29 @@ def update_user(self):
-                        and p.sis_user_id is not null
+            # select all student, tas, and instructors registered for the course
+            user_sql = f"""
+                        with enrollment as (
```
There was [commit](https://github.com/tl-its-umich-edu/my-learning-analytics/commit/8405cd2beedc70e168484683909c9846996b0f99) that made, I don't see that this PR is picking up that changes. The commit was based of this [issue](https://github.com/tl-its-umich-edu/my-learning-analytics/issues/407),those changes need to be included as well

pushyamig commented on dashboard/rules.py:
```diff
@@ -0,0 +1,66 @@
+        )
+        return True
+    except User.DoesNotExist:
+        return False
```
log an error message referencing user x, course 1233 info

pushyamig commented on dashboard/cron.py:
```diff
@@ -128,25 +128,29 @@ def update_user(self):
-                        and c.enrollment_id =  e.enrollment_id
-                        and p.sis_user_id is not null
+            # select all student, tas, and instructors registered for the course
+            user_sql = f"""
```
There is something wrong with this query, I see that if you just have single course the loading the data through cron process works fine, but if you have multiple courses i see this error.
```sqlalchemy.exc.IntegrityError: (_mysql_exceptions.IntegrityError) (1062, "Duplicate entry '17700000000123456-17700000000098765' for key 'PRIMARY'") (Background on this error at: http://sqlalche.me/e/gkpj)```

pushyamig commented on dashboard/rules.py:
```diff
@@ -0,0 +1,66 @@
+rules.add_perm('dashboard.grade_distribution', is_admin_or_enrolled_in_course)
+rules.add_perm('dashboard.update_user_default_selection_for_views', is_admin_or_enrolled_in_course)
+rules.add_perm('dashboard.get_user_default_selection', is_admin_or_enrolled_in_course)
+rules.add_perm('dashboard.assignments', is_admin_or_enrolled_in_course)
```
The rules feature looks fine to me but I couldn't actually test this locally due to load problem error. One concern for me is these rules provide a way to preventing users from accessing other course content. In that case, this PR is not handling it, If not this PR at least another issue must be created stating to handle it from UI, like showing "Permission denied" . I know with rewrite of UI in react it would be a duplicate effort for handling it.

andrew-gardener commented:
@jonespm @pushyamig I've rebased, made some of the fixes you requested, and simplified a few aspects of the way I implemented it.

quick summary:
- simplified the user cron query (https://github.com/tl-its-umich-edu/my-learning-analytics/commit/8405cd2beedc70e168484683909c9846996b0f99 is a much more elegant of doing it than what I did. I just needed to pass through the enrollment type)
- removed the `CANVAS_PRIMARY_AUTH_PROVIDER_ID` setting (no longer needed)
- added the permission checking to the view templates (needed to create new view functions since the rules library doesn't wrap around templates directly very well)
- there is still no permissions within the view, no 403 status checking in ajax requests, or a 403 template at the moment (I think some of this should be addressed in a separate issue after the react rewrite is in place)
